azure-mgmt-resources/src/main/java/com/azure/management/resources/fluentcore/utils/ProviderRegistrationPolicy.java

```diff
@@ -0,0 +1,103 @@
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.management.CloudError;
+import com.azure.core.management.serializer.AzureJacksonAdapter;
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.serializer.SerializerEncoding;
+import com.azure.management.AzureTokenCredential;
+import com.azure.management.RestClient;
+import com.azure.management.RestClientBuilder;
+import com.azure.management.resources.Provider;
+import com.azure.management.resources.fluentcore.arm.ResourceUtils;
+import com.azure.management.resources.implementation.ResourceManager;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A Http Pipeline Policy for automatic provider registration in Azure.
+ */
+public class ProviderRegistrationPolicy implements HttpPipelinePolicy {
+    private final static String MISSING_SUBSCRIPTION_REGISTRATION = "MissingSubscriptionRegistration";
+    private final AzureTokenCredential credential;
+
+    /**
+     * Initialize a provider registration policy with a credential that's authorized
+     * to register the provider.
+     * @param credential the credential for provider registration
+     */
+    public ProviderRegistrationPolicy(AzureTokenCredential credential) {
+        this.credential = credential;
+    }
+
+    private boolean responseSuccessful(HttpResponse response) {
+        return response.getStatusCode() >= 200 && response.getStatusCode() < 300;
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        final HttpRequest request = context.getHttpRequest();
+        return next.process().flatMap(
+            response -> {
+                if (!responseSuccessful(response)) {
+                    HttpResponse bufferedResponse = response.buffer();
+                    return FluxUtil.collectBytesInByteBufferStream(bufferedResponse.getBody()).flatMap(
+                        body -> {
+                            String bodyStr = new String(body, StandardCharsets.UTF_8);
+
+                            AzureJacksonAdapter jacksonAdapter = new AzureJacksonAdapter();
+                            CloudError cloudError;
+                            try {
+                                cloudError = jacksonAdapter.deserialize(bodyStr, CloudError.class, SerializerEncoding.JSON);
+                            } catch (IOException e) {
+                                return Mono.just(bufferedResponse);
+                            }
+
+                            if (cloudError != null && MISSING_SUBSCRIPTION_REGISTRATION.equals(cloudError.getCode())) {
+                                String subscriptionId = ResourceUtils.extractFromResourceId(request.getUrl().getPath(), "subscriptions");
+                                RestClient restClient = new RestClientBuilder()
```

Comment:
Not sure about proxy, or any other configure with origin restClient.

If you are also not sure, add a TODO?

Reply:
Done